### PR TITLE
ENH: Allow removal of (up to two) bad marker coils in read_raw_kit()

### PR DESCRIPTION
**Pull request by JD-Zhu**
@larsoner 

Related to #8981 #8989 

The previous fix deals with missing marker coils when reading in KIT data. However, when all five marker coils are present in the data, sometimes there are bad coils that need to be removed post-hoc. This step needs to happen before the device-head transformation is computed during the call to read_raw_kit().

Removal of bad coils cannot be achieved by simply editing the relevant digitisation file (.elp), as this causes another error downstream in fit_matched_points() due to a mismatch in the number of marker coils between the .elp and .mrk files. The .mrk file cannot be edited as it's not in plain text format. 

The proposed solution here allows a list of bad coils to be passed in when calling read_raw_kit(), so that they can be removed once the marker coil info has been read in from the .elp and .mrk files. Up to two bad coils are allowed, so we can ensure there are at least three marker coils remaining to be used for the tranformation.

### CODE REVIEW

**Comment by larsoner** [comment]
@JD-Zhu if you look at the style run by clicking "Details" you'll see [this](https://github.com/mne-tools/mne-python/actions/runs/7706695438/job/21002662455?pr=12394#step:4:117):
```
mne/io/kit/coreg.py:175:89: E501 Line too long (90 > 88)
mne/io/kit/kit.py:130:19: B006 Do not use mutable data structures for argument defaults
mne/io/kit/kit.py:915:15: B006 Do not use mutable data structures for argument defaults
```
From [CircleCI details](https://app.circleci.com/pipelines/github/mne-tools/mne-python/22838/workflows/f11751f5-6dcc-46c8-835e-866073d0f0fc/jobs/63494) you'll see:
```
RuntimeError: Error documenting RawKIT:
'kit_badcoils'
```
I'll comment more inline but in general when a CI fails if you scroll toward the bottom of the failed/red step it should have helpful info about what went wrong!

File: mne/io/kit/coreg.py [hunk 1]
```diff
@@ -114,7 +114,7 @@ def read_sns(fname):
     return locs
 
 
-def _set_dig_kit(mrk, elp, hsp, eeg):
+def _set_dig_kit(mrk, elp, hsp, bad_coils, eeg):
```
Comment (larsoner) [inline mne/io/kit/coreg.py:117]: To make sure we don't mess anything up anywhere (including in [mne-kit-gui](https://github.com/mne-tools/mne-kit-gui)), we use `*,` to enforce that `bad_coils` must be passed as a keyword argument, not by position:
```suggestion
def _set_dig_kit(mrk, elp, hsp, eeg, *, bad_coils=()):
```
Also let's put the new kwarg at the end, and make the default backward compatible

File: mne/io/kit/coreg.py [hunk 2]
```diff
@@ -133,6 +133,9 @@ def _set_dig_kit(mrk, elp, hsp, eeg):
         Digitizer head shape points, or path to head shape file. If more
         than 10`000 points are in the head shape, they are automatically
         decimated.
+    bad_coils : list | []
```
Comment (larsoner) [inline mne/io/kit/coreg.py:136]: `[]` is a type of list. And we often use list/tuple interchangeably so we can just say
```suggestion
    bad_coils : list
```

File: mne/io/kit/coreg.py [hunk 3]
```diff
@@ -167,12 +170,22 @@ def _set_dig_kit(mrk, elp, hsp, eeg):
                 "%s." % (elp, elp_points.shape)
             )
         elp = elp_points
-    elif len(elp) not in (6, 7, 8):
+        if len(bad_coils) > 0:
+            elp = np.delete(elp, np.add(bad_coils, 3), 0)
```
Comment (larsoner) [inline mne/io/kit/coreg.py:174]: Makes it explicit that it casts to `ndarray` then just uses standard `__add__`
```suggestion
            elp = np.delete(elp, np.array(bad_coils) + 3, 0)
```

File: mne/io/kit/kit.py [hunk 4]
```diff
@@ -43,7 +43,7 @@
 INT32 = "<i4"
 
 
-def _call_digitization(info, mrk, elp, hsp, kit_info):
+def _call_digitization(info, mrk, elp, hsp, bad_coils, kit_info):
```
Comment (larsoner) [inline mne/io/kit/kit.py:46]:
```suggestion
def _call_digitization(info, mrk, elp, hsp, kit_info, *, bad_coils=()):
```

File: mne/io/kit/kit.py [hunk 5]
```diff
@@ -62,7 +62,7 @@ def _call_digitization(info, mrk, elp, hsp, kit_info):
     if mrk is not None and elp is not None and hsp is not None:
         with info._unlock():
             info["dig"], info["dev_head_t"], info["hpi_results"] = _set_dig_kit(
-                mrk, elp, hsp, kit_info["eeg_dig"]
+                mrk, elp, hsp, bad_coils, kit_info["eeg_dig"]
```
Comment (larsoner) [inline mne/io/kit/kit.py:65]:
```suggestion
                mrk, elp, hsp, kit_info["eeg_dig"], bad_coils=bad_coils,
```

File: mne/io/kit/kit.py [hunk 6]
```diff
@@ -91,6 +91,7 @@ class RawKIT(BaseRaw):
     %(kit_mrk)s
     %(kit_elp)s
     %(kit_hsp)s
+    %(kit_badcoils)s
```
Comment (larsoner) [inline mne/io/kit/kit.py:94]: This will need an entry in `mne/utils/docs.py`

File: mne/io/kit/kit.py [hunk 7]
```diff
@@ -126,6 +127,7 @@ def __init__(
         mrk=None,
         elp=None,
         hsp=None,
+        bad_coils=[],
```
Comment (larsoner) [inline mne/io/kit/kit.py:130]:
```suggestion
        *,
        bad_coils=(),
```

File: mne/io/kit/kit.py [hunk 8]
```diff
@@ -905,6 +912,7 @@ def read_raw_kit(
     mrk=None,
     elp=None,
     hsp=None,
+    bad_coils=[],
```
Comment (larsoner) [inline mne/io/kit/kit.py:915]: This one is actually important. If it's not there, it means someone could have passed `mrk, elp, hsp, stim, ...` and now their behavior changes. So we could do this:
```suggestion
    *,
    bad_coils=[],
```
Even safer would be to add this parameter last in the list, after `standardize_names` (with a `*,` preceding it). That's probably better even though this seems like a reasonable place to have the `*,`.

**Review by larsoner — COMMENTED** [review]
Looks like a good start! I think a simple little test would be to look through our existing KIT datasets to see if one of them has enough coils where you could remove one or two of the markers using `bad_coils` and ensure the `dev_head_t` that gets computed is similar (but not identical!). Let me know if anything is unclear or you want help with any of this!

**Comment by drammock** [comment]
> when all five marker coils are present in the data, sometimes there are bad coils that need to be removed post-hoc. This step needs to happen before the device-head transformation is computed during the call to read_raw_kit(). [...] The proposed solution here allows a list of bad coils to be passed in when calling read_raw_kit(), so that they can be removed once the marker coil info has been read in from the .elp and .mrk files.

Perhaps this is a naive question, but does this mean that users need to call `read_raw_kit()` twice?  (once to load all 5 coils and *notice* that some are bad, and a second time to *exclude* the bad ones)?  Would it be possible to load once, notice the bad coils, then call a new method (`RawKIT.remove_marker_coils(bads=[...])` or so) that would re-compute the dev_head_t in-place?  If so, that seems like a preferable / more intuitive user experience to me, and also avoids doing the I/O twice (which is usually slow).

**Comment by JD-Zhu** [comment]
@larsoner Thanks so much for the detailed comments and suggested changes! Is there a way to accept all the suggestions or should I just click "Commit suggestion" for each one? Thanks.

**Comment by JD-Zhu** [comment]
> > when all five marker coils are present in the data, sometimes there are bad coils that need to be removed post-hoc. This step needs to happen before the device-head transformation is computed during the call to read_raw_kit(). [...] The proposed solution here allows a list of bad coils to be passed in when calling read_raw_kit(), so that they can be removed once the marker coil info has been read in from the .elp and .mrk files.
> 
> Perhaps this is a naive question, but does this mean that users need to call `read_raw_kit()` twice? (once to load all 5 coils and _notice_ that some are bad, and a second time to _exclude_ the bad ones)? Would it be possible to load once, notice the bad coils, then call a new method (`RawKIT.remove_marker_coils(bads=[...])` or so) that would re-compute the dev_head_t in-place? If so, that seems like a preferable / more intuitive user experience to me, and also avoids doing the I/O twice (which is usually slow).

Thanks for the comment. In this solution, users only need to call read_raw_kit() once. The bad coils are usually known beforehand (most likely written down during data acquisition), so they can be passed into read_raw_kit() straight away. That said, I think having an option to remove bad coils and re-compute the dev_head_t (as well as updating 'dig' and 'hpi_results'?) for raw data that have already been read in (as per your suggestion) is also not a bad idea.

**Comment by larsoner** [comment]
> Is there a way to accept all the suggestions ("Add suggestion to batch" is grayed out) or should I just click "Commit suggestion" for each one? Thanks.

On the "Files changed" tab of this page (which links [here](https://github.com/mne-tools/mne-python/pull/12394/files)) yes you can "add to batch" then commit multiple changes at once. But at least a couple of my suggestions were more ideas that you'd have to push changes for anyway (or one suggestion that applies in multiple places in principle) so probably easier just to push a commit!

**Comment by JD-Zhu** [comment]
> Looks like a good start! I think a simple little test would be to look through our existing MNE-Python KIT test datasets to see if one of them has enough coils where you could remove one or two of the markers using `bad_coils` and ensure the `dev_head_t` that gets computed is similar (but not identical!). Let me know if anything is unclear or you want help with any of this!

Sounds like a good idea! Will try to test this once I can figure out the virtual environments (which always seem to get me in trouble). 
Alternatively, is there a way to call functions in the dev version I'm working on rather than the stable version I have installed (e.g. maybe specify the path when importing?)

**Comment by larsoner** [comment]
> Alternatively, is there a way to call functions in the dev version I'm working on rather than the stable version I have installed (e.g. maybe specify the path when importing?)

In theory if you are in the `mne-python/` directory, having `mne-python/mne/` directly beneath it should make it so that `import mne` imports from that directory. Looks like that's the case:
```
~$ pip uninstall -q mne
Proceed (Y/n)? y
larsoner:~$ pip install mne
Collecting mne
  Downloading mne-1.6.1-py3-none-any.whl.metadata (13 kB)
Successfully installed mne-1.6.1
~$ python -c "import mne; print(mne)"
<module 'mne' from '/home/larsoner/python/virtualenvs/base/lib/python3.11/site-packages/mne/__init__.py'>
~$ cd python/mne-python
~/python/mne-python$ python -c "import mne; print(mne)"
<module 'mne' from '/home/larsoner/python/mne-python/mne/__init__.py'>
~/python/mne-python$ pytest mne/io/kit/tests/test_kit.py 
=============================================================================== test session starts ================================================================================
platform linux -- Python 3.11.6, pytest-8.0.0, pluggy-1.3.0
PyQt6 6.6.0 -- Qt runtime 6.6.0 -- Qt compiled 6.6.0
MNE 1.6.1 -- /home/larsoner/python/mne-python/mne
...
```
You can see where `mne` gets imported from. Even though the version is reported incorrectly, the path is correct, and I verified that adding a `raise RuntimeError` to `mne/__init__.py` caused the RuntimeError to happen in the `pytest` call.

**Review by larsoner — APPROVED** [review]

**Comment by larsoner** [comment]
Pushed a few commits to make CIs happy then merged, thanks @JD-Zhu !

**Pull request merged**